source/server/connection_handler_impl.cc

```diff
@@ -69,71 +60,141 @@ void ConnectionHandlerImpl::ActiveListener::removeConnection(ActiveConnection& c
       });
 
   // If there is exact address match, return the corresponding listener.
-  if (listener != listeners_.end()) {
-    return listener->second->listener_.get();
+  if (iter != listeners_.end()) {
+    return iter->second.get();
   }
 
   // Otherwise, we need to look for the wild card match, i.e., 0.0.0.0:[address_port].
   // We do not return stopped listeners.
   // TODO(wattli): consolidate with previous search for more efficiency.
-  listener = std::find_if(
+  iter = std::find_if(
       listeners_.begin(), listeners_.end(),
       [&address](const std::pair<Network::Address::InstanceConstSharedPtr, ActiveListenerPtr>& p) {
         return p.second->listener_ != nullptr && p.first->type() == Network::Address::Type::Ip &&
                p.first->ip()->port() == address.ip()->port() && p.first->ip()->isAnyAddress();
       });
-  return (listener != listeners_.end()) ? listener->second->listener_.get() : nullptr;
+  return (iter != listeners_.end()) ? iter->second.get() : nullptr;
+}
+
+void ConnectionHandlerImpl::ActiveSocket::continueFilterChain(bool success) {
+  if (success) {
+    ActiveListener* new_listener = nullptr;
+
+    if (iter_ == accept_filters_.end()) {
+      iter_ = accept_filters_.begin();
+    } else {
+      iter_ = std::next(iter_);
+    }
+
+    for (; iter_ != accept_filters_.end(); iter_++) {
+      Network::FilterStatus status = (*iter_)->onAccept(*this);
+      if (status == Network::FilterStatus::StopIteration) {
+        // The filter is responsible for calling us again at a later time to continue the filter
+        // chain from the next filter.
+        return;
+      }
+    }
+    // Successfully ran all the accept filters.
+
+    // Check if the socket may need to be redirected to another listener.
+    if (!redirected_ && socket_->localAddressReset()) {
+      // Find a listener associated with the original destination address.
+      new_listener = listener_->parent_.findActiveListenerByAddress(*socket_->localAddress());
+    }
+    if (new_listener != nullptr) {
+      // Hands off connections redirecrted by iptables to the listener associated with the
+      // original destination address. Pass 'redirected' as true to prevent further redirection.
+      new_listener->onAccept(std::move(socket_), true);
+    } else {
+      // Create a new connection on this listener.
+      listener_->newConnection(std::move(socket_));
+    }
+  }
+
+  // Filter execution concluded, clear state.
+  iter_ = accept_filters_.end();
+  ActiveSocketPtr removed = removeFromList(listener_->sockets_);
+  listener_->parent_.dispatcher_.deferredDelete(std::move(removed));
+}
+
+void ConnectionHandlerImpl::ActiveListener::onAccept(Network::AcceptedSocketPtr&& socket,
+                                                     bool redirected) {
+  Network::Address::InstanceConstSharedPtr local_address = socket->localAddress();
+  ActiveSocket* active_socket(new ActiveSocket(*this, std::move(socket), redirected));
```

Comment:
std::unique_ptr (please no naked pointers anywhere)

Reply:
Fixed by linking the unique_ptr only if filter iteration does not conclude on the first call to continueFIlterChain().